### PR TITLE
disable copy constructor and copy assignment

### PR DESCRIPTION
**Pull request by allenhan2**
Due to the way those types are initialized, it is not safe to allow the copy constructor or copy-assignment operator on the kv table type. 

## Change Description

- disable copy constructor and assignment
- disable move constructor and assignment 

## API Changes
- [ ] API Changes

## Documentation Additions
- [x] Documentation Additions

Doc team, please make sure our examples were not using these accidentally, thanks.

### CODE REVIEW

File: libraries/eosiolib/contracts/eosio/key_value.hpp [hunk 1]
```diff
@@ -503,6 +503,11 @@ class table : internal::table_base {
    template<typename K>
    class index;
 
+   table(const table&) = delete;                         // disbale copy constructor
```
Comment (b1bart) [inline libraries/eosiolib/contracts/eosio/key_value.hpp:506]: spelling error in disable

File: libraries/eosiolib/contracts/eosio/key_value.hpp [hunk 2]
```diff
@@ -503,6 +503,11 @@ class table : internal::table_base {
...
+   table(table&&) = default;                             // default move constructor
+   table& operator=(const table&) = delete;              // disbale assginment copy operator
```
Comment (b1bart) [inline libraries/eosiolib/contracts/eosio/key_value.hpp:508]: spelling error in assignment

**Review by b1bart — CHANGES_REQUESTED** [review]

**Review by larryk85 — COMMENTED** [review]
What is currently unsafe about it?

**Comment by tbfleming** [comment]
It holds internal pointers to index members

**Comment by tbfleming** [comment]
Which, BTW, is also incompatible with move. That should also be disabled.

**Comment by allenhan2** [comment]
> Which, BTW, is also incompatible with move. That should also be disabled.

Could you please elaborate a scenario for problem of move? thanks!

**Comment by tbfleming** [comment]
Simplification of what's going on:

```
struct foo {
    int a = 0;
    int* p;

    foo(): p{&a} {}
};
```

**Comment by allenhan2** [comment]
It is a great example, but could it happen in the table class? disable move will make hard for utility functions.

**Comment by tbfleming** [comment]
Like I said above, kv_table does this. It's part of how it keeps track of indexes.

**Comment by allenhan2** [comment]
okay, then, will disable the move.

**Review by tbfleming — APPROVED** [review]

**Review by larryk85 — APPROVED** [review]

**Comment by allenhan2** [comment]
no permission to merge, submitted tick for permission, https://blockone.freshservice.com/support/tickets/15178

**Comment by allenhan2** [comment]
@b1bart could you give another review, the merge is blocked by the request change, thanks,

**Pull request merged**